README.md
# sicinfo-model

CHANGELOG.md
## sicinfo-model

### [0.2.0] [2018-06-21] Moreira
#### corrige 
- save method

#### inclui
- delete method

#### altera
- static send method
- static fetch method
- static sync method

### [0.1.2] [2018-06-21] Moreira
#### incluir
- save no model

### [0.1.2] [2018-06-21] Moreira
#### corrige
- Sync.url sequencia para monta url: param seguido de query

### [0.1.1] [2018-05-27] Moreira
#### altera Model
- inclui
```
get attributes()
```


### [0.1.0] [2018-05-15] Moreira
- melhorias para Collection e Model
- Sync.parse recursivo


### [0.0.7] 2018-04-02
- corrige sync.url

### [0.0.6] 2018-03-08
- corrige sync.url

### [0.0.5] 2018-03-08
- corrige Sync.fetch, Sync.sync

### [0.0.4] 2018-03-08
- corrige model.fetch

### [0.0.3] 2018-03-08
- corrige sync.url

### [0.0.2] 2018-03-07
- inclui funcionalidade para compatibilizar com backbone

### [0.0.1] 2018-02-27
- uso experimental citsmart-labti

test/model.test.js
/** 
 * application: sicinfo-model
 * 
 */

const model = require(`../lib/model`)
const tape = require(`tape`)


let count = 1;
const t1 = (label, result, test) => tape(`${count++}.${label}`, t => {

  (assert => t.assert(result === assert, `${result} - ${assert}`))(test());
  t.end();
  console.log('-'.repeat(80))

});

const j0 = JSON.stringify;
const t2 = (label, result, test) => t1(label, j0(result), () => j0(test()))

// require('./model-sync.test')(tape, model, t1, t2, j0)
require('./model-model.test')(tape, model, t1, t2, j0)


// console.log('inicia -->\n')


// // // Sync
// // t1(`1 sync constructor`, true, () => !!new Sync(`a`));
// // t1(`2 sync constructor`, true, () => !!new Sync({ 'url': `a` }));
// // t1(`3 sync constructor`, true, () => !!new Sync());
// // t1(`4 sync constructor`, true, () => !!new Sync(1));
// // t1(`5 sync constructor`, true, () => !!new Sync(true));
// // t1(`6 sync constructor`, true, () => !!new Sync({}));
// // t1(`7 sync constructor`, true, () => !!new Sync(false));
// // t1(`8 sync constructor`, true, () => !!new Sync(1 == 1));
// // t1(`9 sync constructor`, true, () => !!new Sync(1 == 2));

// // console.log('='.repeat(80))

// // t1(`1 sync url`, `a`, () => new Sync(`a`).$url())
// // t1(`2 sync url`, `1`, () => new Sync(1).$url())
// // t1(`3 sync url`, `true`, () => new Sync(true).$url())
// // t1(`4 sync url`, `true`, () => new Sync(1 == 1).$url())
// // t1(`5 sync url`, `false`, () => new Sync(false).$url())
// // t1(`6 sync url`, `false`, () => new Sync(1 == 2).$url())
// // t1(`7 sync url`, undefined, () => new Sync().$url())
// // t1(`8 sync url`, undefined, () => new Sync({}).$url())
// // t1(`9 sync url`, undefined, () => new Sync(null).$url())
// // t1(`10 sync url`, `a`, () => new Sync().$url('a'))
// // t1(`11 sync url`, `1`, () => new Sync().$url(1))
// // t1(`12 sync url`, `true`, () => new Sync().$url(true))
// // t1(`13 sync url`, `true`, () => new Sync().$url(1 == 1))
// // t1(`14 sync url`, `false`, () => new Sync().$url(false))
// // t1(`15 sync url`, `false`, () => new Sync().$url(1 == 2))
// // t1(`16 sync url`, undefined, () => new Sync().$url({}))
// // t1(`17 sync url`, undefined, () => new Sync().$url(null))

// // console.log('='.repeat(80))

// // t2(`1 sync parse`, undefined, () => new Sync().$parse())
// // t2(`1 sync parse`, [], () => new Sync().$parse([]))
// // t2(`2 sync parse`, [[]], () => new Sync().$parse([[]]))
// // {
// //   let a = [{ 'a': 'a', 'b': 1 }, { 'a': 'b', 'b': 2 }]
// //   let b = [['a', 'b'], ['a', 1], ['b', 2]]
// //   let c = new Collection()
// //   let d = new Model()
// //   let e = { 'a': a }
// //   let f = { 'a': b }
// //   let g = { 'a': 'a' }

// //   t2(`3 sync parse`, a, () => new Sync().$parse.call(c, b))
// //   t2(`4 sync parse`, b, () => new Sync().$parse(b))
// //   t2(`5 sync parse`, g, () => new Sync().$parse.call(d, g))
// //   t2(`6 sync parse`, e, () => new Sync().$parse.call(d, f))
// // }

// // console.log('='.repeat(80))

// // t1(`1 model constructor`, true, () => !!new Model());
// // t1(`2 model constructor`, true, () => !!new Model(null, 'aa'));
// // t1(`3 model constructor`, true, () => !!new Model(null, { 'url': 'a' }));
// // t1(`4 model constructor`, true, () => !!new Model({ 'aa': 'aa' }, { 'url': 'a' }));

// // {
// //   let c1 = new Collection();

// //   let a1 = { 'nome': 'jose' }
// //   let a2 = { 'nome': 'maria' }
// //   let a3 = { 'id': 'a', 'nome': 'margarida', 'idade': 35 }

// //   let m1 = new Model()

// //   let m2 = new Model(null, { 'idAttribute': 'a', 'cidPrefix': 'x' })
// //   m2.$set('idade', 25)
// //   m2.$set({ 'nome': 'maria' })

// //   let m3 = new Model(a1, {
// //     'defaults': a2,
// //     'collection': c1
// //   })

// //   let m4 = new Model(a3)
// //   m4.$unset('nome')
// //   m4.$unset({ 'idade': undefined })

// //   let m5 = new Model(a3)
// //   m5.$clear()

// //   t1(`1 model $getId`, undefined, () => m1.$getId());
// //   t1(`2 model $getId`, 'a', () => m4.$getId());

// //   console.log('='.repeat(80))

// //   t1(`1 model idAttribute`, 'id', () => m1.$.idAttribute);
// //   t1(`2 model idAttribute`, 'a', () => m2.$.idAttribute);

// //   console.log('='.repeat(80))

// //   t1(`1 model cid`, 'm_1', () => m1.$.cid);
// //   t1(`2 model cid`, 'x_2', () => m2.$.cid);
// //   t1(`3 model cid`, 'm_3', () => m3.$.cid);

// //   console.log('='.repeat(80))

// //   t2(`1 model defaults`, {}, () => m1.$.defaults);
// //   t2(`2 model defaults`, a2, () => m3.$.defaults);

// //   console.log('='.repeat(80))

// //   t1(`1 model collection`, undefined, () => m1.$.collection);
// //   t1(`2 model collection`, c1, () => m3.$.collection);

// //   console.log('='.repeat(80))

// //   t2(`1 model attributes`, {}, () => m1.$.attributes);
// //   t2(`2 model attributes`, a1, () => m3.$.attributes);
// //   t2('3 model attributes', { 'idade': 25, 'nome': 'maria' }, () => m2.$.attributes)
// //   t2('4 model attributes', { 'id': 'a' }, () => m4.$.attributes)

// //   console.log('='.repeat(80))

// //   t2('1 model $toJSON', { 'idade': 25, 'nome': 'maria' }, () => m2.$toJSON())
// //   t2('2 model $toJSON', {}, () => m5.$toJSON())

// //   console.log('='.repeat(80))

// //   t1(`1 model isNew`, true, () => m1.$isNew())
// //   t1(`2 model isNew`, false, () => m4.$isNew())

// //   console.log('='.repeat(80))

// //   t1(`1 model $hasChanged`, false, () => m1.$hasChanged())
// //   t1(`2 model $hasChanged`, true, () => m2.$hasChanged())
// //   t1(`3 model $hasChanged`, true, () => m2.$hasChanged('nome'))
// //   t1(`4 model $hasChanged`, false, () => m4.$hasChanged('nome'))
// //   t1(`5 model $hasChanged`, false, () => m5.$hasChanged())

// //   console.log('='.repeat(80))

// //   t2(`1 model $changedAttributes`, {}, () => m5.$changedAttributes())

// //   console.log('='.repeat(80))

// //   t1('1 model get', undefined, () => m1.nome);

// //   console.log('='.repeat(80))

// //   t1('1 model has', false, () => m1.$has('nome'));

// // console.log('='.repeat(80))
// // }


// {
//   // let m1 = []

//   // t1(`1 collection constructor`, true, () => !!new Collection());
//   // t1(`2 collection constructor`, true, () => !!new Collection(m1));
//   // console.log('='.repeat(80))

//   console.log('='.repeat(80))

//   let m1 = new Model()
//   let m2 = new Model()
//   let c1 = new Collection([m1, m2])

//   console.log('='.repeat(80))
//   console.log(c1.$)
//   console.log(c1.$.length)
//   c1.$.models.some(m => console.log(m.$))

//   // c1.$_removeModels([m1])

//   console.log('------------------------------------------------')
//   console.log(m1.$)
//   console.log(c1.$)
//   console.log('------------------------------------------------')

//   m1.$set('id','x')
//   // c1.$add(m1)


//   console.log('='.repeat(80))
//   console.log(c1.$)
//   console.log(c1.$.length)
//   c1.$.models.some(m => console.log(m.$))

//   console.log('='.repeat(80))
//   return
// }

// // {
// //   class M extends Model {
// //     constructor(a,o = {}) {
// //       o.idAttribute = 'x'
// //       super(a,o)
// //     }
// //   }

// //   let c1 = new Collection();
// //   let c2 = new Collection(null, {'model': M});

// //   t1('1 collection $modelId', undefined, () => c1.$modelId())
// //   t1('2 collection $modelId', undefined, () => c1.$modelId({}))
// //   t1('3 collection $modelId', 'a', () => c1.$modelId({ 'id': 'a' }))
// //   t1('4 collection $modelId', 'a', () => c2.$modelId({ 'x': 'a' }))
// //   console.log('='.repeat(80))
// // }

// {
//   let m2 = new Model()
//   let c2 = new Collection([])

//   // let m3 = new Model()
//   // let c3 = new Collection([m3]).$removeReference(m3)

//   // t2('1 collection $removeReference', c2, () => m2.$.collection)

//   console.log('='.repeat(80))
// }


test/model.test.1.js
/** 
 * application: sicinfo-model
 * 
 */

const { Sync, Model, Collection } = require(`../lib/model`)
const tape = require(`tape`)

let count = 1;
const t1 = (label, result, test) => tape(`${count++}.${label}`, t => {

  (assert => t.assert(result === assert, `${result} - ${assert}`))(test());
  t.end();
  console.log('-'.repeat(80))

});

const j0 = JSON.stringify;
const t2 = (label, result, test) => t1(label, j0(result), () => j0(test()))

console.log('inicia -->\n')


// // Sync
// t1(`1 sync constructor`, true, () => !!new Sync(`a`));
// t1(`2 sync constructor`, true, () => !!new Sync({ 'url': `a` }));
// t1(`3 sync constructor`, true, () => !!new Sync());
// t1(`4 sync constructor`, true, () => !!new Sync(1));
// t1(`5 sync constructor`, true, () => !!new Sync(true));
// t1(`6 sync constructor`, true, () => !!new Sync({}));
// t1(`7 sync constructor`, true, () => !!new Sync(false));
// t1(`8 sync constructor`, true, () => !!new Sync(1 == 1));
// t1(`9 sync constructor`, true, () => !!new Sync(1 == 2));

// console.log('='.repeat(80))

// t1(`1 sync url`, `a`, () => new Sync(`a`).$url())
// t1(`2 sync url`, `1`, () => new Sync(1).$url())
// t1(`3 sync url`, `true`, () => new Sync(true).$url())
// t1(`4 sync url`, `true`, () => new Sync(1 == 1).$url())
// t1(`5 sync url`, `false`, () => new Sync(false).$url())
// t1(`6 sync url`, `false`, () => new Sync(1 == 2).$url())
// t1(`7 sync url`, undefined, () => new Sync().$url())
// t1(`8 sync url`, undefined, () => new Sync({}).$url())
// t1(`9 sync url`, undefined, () => new Sync(null).$url())
// t1(`10 sync url`, `a`, () => new Sync().$url('a'))
// t1(`11 sync url`, `1`, () => new Sync().$url(1))
// t1(`12 sync url`, `true`, () => new Sync().$url(true))
// t1(`13 sync url`, `true`, () => new Sync().$url(1 == 1))
// t1(`14 sync url`, `false`, () => new Sync().$url(false))
// t1(`15 sync url`, `false`, () => new Sync().$url(1 == 2))
// t1(`16 sync url`, undefined, () => new Sync().$url({}))
// t1(`17 sync url`, undefined, () => new Sync().$url(null))

// console.log('='.repeat(80))

// t2(`1 sync parse`, undefined, () => new Sync().$parse())
// t2(`1 sync parse`, [], () => new Sync().$parse([]))
// t2(`2 sync parse`, [[]], () => new Sync().$parse([[]]))
// {
//   let a = [{ 'a': 'a', 'b': 1 }, { 'a': 'b', 'b': 2 }]
//   let b = [['a', 'b'], ['a', 1], ['b', 2]]
//   let c = { 'models': [] }
//   let d = { 'cid': 'a' }
//   let e = { 'a': a }
//   let f = { 'a': b }
//   let g = { 'a': 'a' }

//   t2(`3 sync parse`, a, () => new Sync().$parse.call(c, b))
//   t2(`4 sync parse`, b, () => new Sync().$parse(b))
//   t2(`5 sync parse`, g, () => new Sync().$parse.call(d, g))
//   t2(`6 sync parse`, e, () => new Sync().$parse.call(d, f))
// }

// console.log('='.repeat(80))

// t1(`1 model constructor`, true, () => !!new Model());
// t1(`2 model constructor`, true, () => !!new Model(null, 'aa'));
// t1(`3 model constructor`, true, () => !!new Model(null, { 'url': 'a' }));
// t1(`4 model constructor`, true, () => !!new Model({ 'aa': 'aa' }, { 'url': 'a' }));

// {
//   let c1 = new Collection();

//   let a1 = { 'nome': 'jose' }
//   let a2 = { 'nome': 'maria' }
//   let a3 = { 'id': 'a', 'nome': 'margarida', 'idade': 35 }

//   let m1 = new Model()

//   let m2 = new Model(null, { 'idAttribute': 'a', 'cidPrefix': 'x' })
//   m2.$set('idade', 25)
//   m2.$set({ 'nome': 'maria' })

//   let m3 = new Model(a1, {
//     'defaults': a2,
//     'collection': c1
//   })

//   let m4 = new Model(a3)
//   m4.$unset('nome')
//   m4.$unset({ 'idade': undefined })

//   let m5 = new Model(a3)
//   m5.$clear()

//   t1(`1 model $getId`, undefined, () => m1.$getId());
//   t1(`2 model $getId`, 'a', () => m4.$getId());

//   console.log('='.repeat(80))

//   t1(`1 model idAttribute`, 'id', () => m1.$.idAttribute);
//   t1(`2 model idAttribute`, 'a', () => m2.$.idAttribute);

//   console.log('='.repeat(80))

//   t1(`1 model cid`, 'm_1', () => m1.$.cid);
//   t1(`2 model cid`, 'x_2', () => m2.$.cid);
//   t1(`3 model cid`, 'm_3', () => m3.$.cid);

//   console.log('='.repeat(80))

//   t2(`1 model defaults`, {}, () => m1.$.defaults);
//   t2(`2 model defaults`, a2, () => m3.$.defaults);

//   console.log('='.repeat(80))

//   t1(`1 model collection`, undefined, () => m1.$.collection);
//   t1(`2 model collection`, c1, () => m3.$.collection);

//   console.log('='.repeat(80))

//   t2(`1 model attributes`, {}, () => m1.$.attributes);
//   t2(`2 model attributes`, a1, () => m3.$.attributes);
//   t2('3 model attributes', { 'idade': 25, 'nome': 'maria' }, () => m2.$.attributes)
//   t2('4 model attributes', { 'id': 'a' }, () => m4.$.attributes)

//   console.log('='.repeat(80))

//   t2('1 model $toJSON', { 'idade': 25, 'nome': 'maria' }, () => m2.$toJSON())
//   t2('2 model $toJSON', {}, () => m5.$toJSON())

//   console.log('='.repeat(80))

//   t1(`1 model isNew`, true, () => m1.$isNew())
//   t1(`2 model isNew`, false, () => m4.$isNew())

//   console.log('='.repeat(80))

//   t1(`1 model $hasChanged`, false, () => m1.$hasChanged())
//   t1(`2 model $hasChanged`, true, () => m2.$hasChanged())
//   t1(`3 model $hasChanged`, true, () => m2.$hasChanged('nome'))
//   t1(`4 model $hasChanged`, false, () => m4.$hasChanged('nome'))
//   t1(`5 model $hasChanged`, false, () => m5.$hasChanged())

//   console.log('='.repeat(80))

//   t2(`1 model $changedAttributes`, {}, () => m5.$changedAttributes())

//   console.log('='.repeat(80))

//   t1('1 model get', undefined, () => m1.nome);

//   console.log('='.repeat(80))

//   t1('1 model has', false, () => m1.$has('nome'));

// console.log('='.repeat(80))
// }


// {
//   let m1 = []

//   t1(`1 collection constructor`, true, () => !!new Collection());
//   console.log('='.repeat(80))
// }

// {
//   class M extends Model {
//     constructor(a,o = {}) {
//       o.idAttribute = 'x'
//       super(a,o)
//     }
//   }

//   let c1 = new Collection();
//   let c2 = new Collection(null, {'model': M});

//   t1('1 collection $modelId', undefined, () => c1.$modelId())
//   t1('2 collection $modelId', undefined, () => c1.$modelId({}))
//   t1('3 collection $modelId', 'a', () => c1.$modelId({ 'id': 'a' }))
//   t1('4 collection $modelId', 'a', () => c2.$modelId({ 'x': 'a' }))
//   console.log('='.repeat(80))
// }

// {
//   let m2 = new Model()
//   let c2 = new Collection([m2])

//   let m3 = new Model()
//   let c3 = new Collection([m3]).$removeReference(m3)

//   t2('1 collection $removeReference', c2, () => m2.$.collection)

//   console.log('='.repeat(80))
// }






















// // t1(`model`, `m_1`, () => new Model().cid);

// // t1(`model`, `x_2`, () => new Model(null, { `cidPrefix`: `x` }).cid);

// // t1(`model`, `m_3`, () => new Model().cid);

// // t1(`model`, `s`, () => new Model({ `a`: `s` }).get(`a`));

// // t1(`model`, JSON.stringify({ `a`: `a`, `b`: `b` }), () => {

// //   let model = new Model();

// //   model.set(`a`, `a`)
// //   model.set({ `b`: `b` })

// //   return JSON.stringify(model.toJSON())

// // })

// // t1(`model`, `z`, () => {

// //   let model = new Model();

// //   model.set(`id`, `z`)

// //   return model.getId()

// // })


// // t1(`model isNew`, true, () => {

// //   let model1 = new Model(null, { `idAttribute`: `x` });

// //   let model2 = new Model({ `id`: `z` })

// //   return !model2.isNew() && model1.isNew()

// // })

// // t1(`model unset`, true, () => {

// //   let model = new Model({ `x`: `x` });

// //   model.unset(`x`, `z`)

// //   return !model.has(`x`)

// // })

// // // t1(`model fetch`, `test`, async () => {

// // //   let model = new Model({`url`: `package.json`});

// // //   await model.fetch();

// // //   console.log(model)


// // //   return model.get(`name`)

// // // })

// // t1(`model unset`, true, () => {

// //   let model = new Model({ `x`: `x` });

// //   model.unset(`x`, `z`)

// //   return !model.has(`x`)

// // })



// //======================================================================

// // t1(`collection constructor`, true, () => {

// //   let c, m, r = []

// //   m = new Model()

// //   c = new Collection(m)

// //   r.push(c instanceof Collection)

// //   c = new Collection([])

// //   r.push(c instanceof Collection)

// //   c = new Collection([m])

// //   r.push(c instanceof Collection)

// //   c = new Collection([m, new Model()])

// //   r.push(c instanceof Collection)

// //   r.push(c.length == c.models.length)

// //   c = new Collection(null, { `url`: `teste` })

// //   r.push(c instanceof Collection)

// //   return r.every(a => a)

// // })

// // t1(`collection model`, true, () => {

// //   class Teste extends Model {

// //   }

// //   let c1 = new (new Collection(null, { `model`: Teste })).model()
// //   let c2 = new (new Collection()).model()




// //   return c1 instanceof Teste && !(c2 instanceof Teste);

// // })

// // t1(`collection url`, true, () => {

// //   class C extends Collection {
// //     constructor(models, options) {
// //       super(models, Object.assign({ `url`: `teste` }, options))
// //     }

// //     url() {
// //       return `${super.$url()}/aaa`
// //     }
// //   }

// //   return `teste/aaa` == new C().$url();

// // })

// t1(`collection get / set`, true, () => {

//   let r = []
//   let m1 = new Model({`nome`:`jose`})

//   let c1 = new Collection(m1)

//   // r.push(!c1.set())
//   // r.push(!c1.set(null))
//   // r.push(!!c1.set(new Model()))

//   r.push(!!c1.set([{`nome`:`maria`}, new Model()]))

//   r.push(!!c1.set([{`nome`:`maria`}, new Model()]))

//   r.push(m1 == c1.get(m1))

//   console.log(`c1.get(m1)`, c1.get(m1))
//   console.log(c1.length)
//   console.log(c1.models.length)


//   // console.log(c1.models)
//   // c1.set(new )


//   return r.every(e => e)

// })



test/model-sync.test.js
/** 
 * application: sicinfo-model
 * 
 */

module.exports = (tape, { Sync }, t1, t2) => {

  console.log('inicia -->\n')

  // Sync
  {

    let sync = new Sync()

    t1(`1 sync constructor`, true, () => sync instanceof Sync)

    t1(`2 sync constructor`, 'object', () => {

      let { $ } = sync

      return typeof $
    })
  }

  {

    let sync = new Sync()
    let $this = { '$': { 'url': 'aaaa' } }
    let opt = { 'url': 'bbbb' }
    let par = { 'url': 'cccc', 'param': '1' }

    t1(`1 sync url`, '', () => sync.url())
    t1(`2 sync url`, 'aaaa', () => sync.url.call($this))
    t1(`3 sync url`, 'bbbb', () => sync.url(opt))
    t1(`4 sync url`, 'bbbb', () => sync.url.call($this, opt))
    t1(`5 sync url`, 'cccc/1', () => sync.url(par))

  }

  {
    let a = [{ 'a': 'a', 'b': 1 }, { 'a': 'b', 'b': 2 }]
    let b = [['a', 'b'], ['a', 1], ['b', 2]]

    t1(`1 sync parse`, undefined, () => Sync.parse())
    t1(`2 sync parse`, 'a', () => Sync.parse('a'))
    t2(`3 sync parse`, [], () => Sync.parse([]))
    t2(`4 sync parse`, a, () => Sync.parse(a))
    t2(`5 sync parse`, a, () => Sync.parse(b))
  }
}

lib/model.js
/**
 * applicatiion: lib/fetch
 * 
 * [0.1.0] [2018-05-15T10:20]
 * 
 * [0.0.2] 2018-03-07
 * 
 * [0.0.1] 2018-02-27
 * - experimental
 * 
 * ref:
 * https://www.tutorialspoint.com/backbonejs/backbonejs_collection.htm
 * 
 */

((root, factory) => {

  const uniqueid = Symbol(); root[uniqueid] = 0;

  if ('undefined' != typeof (define))
    define([], () => factory(root, a => `${a}_${++root[uniqueid]}`));

  else if ('undefined' != typeof (process) && process.versions && process.versions.node)
    module.exports = factory(root, a => `${a}_${++root[uniqueid]}`);

})(this || self || window, (root, uniqueId) => {


  const assign = Object.assign;
  const clone = (...a) => assign({}, ...a);
  const fetch = root && root.fetch;
  const isArray = Array.isArray;

  class Sync {

    constructor(options) {
      Object.defineProperty(this, '$', { 'value': {}, 'writable': false });
    }

    url(options = {}) {

      let { url, param, query } = options;

      if (undefined === url) url = this.$.url || '';

      if (param) {
        if (isArray(param)) param = `${param.join('/')}`;
        if (!param.startsWith('/')) param = `/${param}`;
        url = `${url}${param}`;
      }

      if (query) {
        if ('object' === typeof (query)) {
          const esc = encodeURIComponent;
          query = Object.keys(query).map(key => `${esc(key)}=${esc(query[key])}`);
        }
        if (isArray(query)) query = query.join('&');
        if (!query.startsWith('?')) query = `?${query}`;
        url = `${url}${query}`;
      }

      return url;
    }

    static parse(arg) {

      if (isArray(arg) && arg.length > 1 &&
        isArray(arg[0]) && isArray(arg[1]) &&
        arg[0].length == arg[1].length &&
        arg[0].every(key => 'string' === typeof (key))) {

        const header = arg[0], { length } = header;

        return arg.slice(1).map(values => {
          const resp = {};
          for (let index = 0; index < length; index++) {
            resp[header[index]] = Sync.parse(values[index]);
          }
          return resp;
        });

      }

      return arg;

    }

    static sync(url, options) {

      if (!('headers' in options.syncOptions)) {
        options.syncOptions.headers = new Headers();
      }

      options.syncOptions.headers.set('content-type', 'application/json');

      return new Promise((resolve, reject) => {
        fetch(url, options.syncOptions)
          .then(response => {
            if (response.ok) response.json().then(attributes => resolve({ attributes, response, options }))
            else reject(response);
          })
          .catch(err => reject(err))
      })
    }

    static send(method, options) {
      options.syncOptions = { method, 'body': JSON.stringify(this.attributes) };
      return Sync.sync(this.url(options), options);
    }

    static fetch(options) {
      options.syncOptions = { 'method': 'GET' };
      return Sync.sync(this.url(options), options);
    }

  }

  class Model extends Sync {

    constructor(attrs, options = {}) {
      super(options);
      const { idAttribute, defaults, cidPrefix, collection } = options;
      if (collection) this.$.collection = collection;
      this.$.idAttribute = idAttribute || 'id';
      this.$.defaults = defaults || {};
      this.$.cid = uniqueId(cidPrefix || 'm');
      this.$.attributes = {};
      this.set(clone(this.$.defaults, attrs), options)
    }

    get id() {
      return this.$.id
    }

    get attributes() {
      return this.$.attributes
    }

    get collection() {
      return this.$.collection || undefined;
    }

    get idAttribute() {
      return this.$.idAttribute;
    }

    get isNew() {
      return !this.has(this.$.idAttribute)
    }

    has(attr) {
      return attr in this.$.attributes;
    }

    get(attr) {
      return this.$.attributes[attr]
    }

    set(attrs, val, options = {}) {

      if ('object' === typeof (attrs)) val && (options = val);
      else (key => (attrs = {})[key] = val)(attrs);

      if (!this.validate(attrs, options)) return false

      this.$.previousAttributes = this.$.attributes;

      const current = this.$.attributes
      const unset = options.unset

      Object.keys(attrs).some(attr => {
        if (unset) { if (attr in current) delete current[attr] }
        else if (!(attr in current && current[attr] === val)) current[attr] = attrs[attr]
      })

      if (this.$.idAttribute in attrs) this.$.id = this.get(this.$.idAttribute);

      return this;
    }

    unset(attr, options) {
      return this.set(attr, undefined, clone(options, { 'unset': true }));
    }

    clear(options) {
      const attrs = {};
      Object.keys(this.$.attributes).some(attr => { attrs[attr] = undefined });
      return this.set(attrs, clone(options, { 'unset': true }));
    }

    toJSON() {
      return clone(this.$.attributes);
    }

    hasChanged(attr) {
      const changed = this.changedAttributes();
      return undefined === attr ? !!Object.keys(changed).length : attr in changed
    }

    changedAttributes() {
      const { previousAttributes, attributes } = this.$;
      const changed = {};

      Object.keys(attributes)
        .filter(attr => !(attr in previousAttributes && previousAttributes[attr] === attributes[attr]))
        .some(attr => { changed[attr] = attributes[attr] });

      return changed;
    }

    parse(resp, options) {

      if (resp && 'object' === typeof (resp)) {
        Object.keys(resp).some(attr => { resp[attr] = Sync.parse(resp[attr]) });
      }

      return resp
    }

    url(options = {}) {
      const { param } = options;

      if (!param && !this.isNew) {
        options.param = `${this.id}`;
      }

      return super.url(options);
    }

    fetch(options) {
      const model = this;
      const { idAttribute } = model.$;

      options = assign({ 'parse': true }, options);
      idAttribute in options && model.set(idAttribute, options[idAttribute]);

      return new Promise((resolve, reject) => {

        if (model.isNew) {
          reject({ 'error': 'modelo sem identificador' });
        }

        else {
          Sync.fetch.call(model, options)
            .then(({ attributes, response, options }) => {
              model.set(options.parse ? model.parse(attributes, options) : attributes, options)
              resolve({ model, response, options })
            });
        }

      })
    }

    delete(attrs = null, val, options = {}) {
      if (attrs) {
        if ('object' === typeof (attrs)) val && (options = val);
        else (key => (attrs = {})[key] = val)(attrs);
      }
      return this.save(attrs, Object.assign({ 'delete': true }, options))
    }

    save(attrs, val, options = {}) {

      if ('object' === typeof (attrs)) val && (options = val);
      else (key => (attrs = {})[key] = val)(attrs);
      options = Object.assign({ 'parse': true }, options);

      return new Promise((resolve, reject) => {

        if (attrs && !this.set(attrs, options)) reject(this);

        else {
          Sync.send.call(this, this.isNew ? 'POST' : (options.delete ? 'DELETE' : 'PUT'), options)
        }

      });




    }

    validate(attrs, options) {
      return true;
    }

  }

  class Collection extends Sync {

    constructor(models, options = {}) {
      const { model } = options;
      super(options);
      this.$.model = model || Model;
      this._reset();
      if (models) this.reset(models, options);
    }

    get model() {
      return this.$.model
    }

    get models() {
      return this.$.models
    }

    get length() {
      return this.models.length
    }

    reset(models, options) {
      options = clone(options)

      this.models.some(model => { this._removeReference(model, options) });
      this._reset();

      return this.add(models, options);
    }

    add(models, options) {
      return this.set(models, clone(options, { 'add': true, 'remove': false }));
    }

    remove(models, options) {
      options = clone(options);
      const singular = !isArray(models);
      if (singular) models = [models];

      const removed = this._removeModels(models, options);

      return singular ? removed[0] : removed;
    }

    // collecion.set
    set(models, options) {

      if (undefined === models) return
      options = assign({ 'add': true, 'remove': true, 'merge': true }, options);

      if (options.parse && !this._isModel(models)) {
        models = this.parse(models, options) || [];
      }

      const { add, merge, remove } = options;
      const replace = add && remove;
      const set = [], toAdd = [], toMerge = [], modelMap = {};
      const singular = !isArray(models);

      if (singular) models = [models]

      models.some((model, i) => {

        const existing = this.get(model);

        // console.log('existing', existing)
        // console.log('add', add)
        // console.log('remove', remove)
        // console.log('merge  ----------> ', merge)

        if (existing) {

          // console.log('model  ----------> ', model.$)
          // console.log('existing  -------> ', existing.$)
          // console.log('model !== existing ', model !== existing)

          if (merge && (model.$ && model.$.cid) != existing.$.cid) {
            const attrs = this._isModel(model) ? model.$.attributes : model;
            if (options.parse) attrs = existing.parse(attrs, options);
            existing.set(attrs, options);
            toMerge.push(existing);
          }

          // console.log('!modelMap[existing.$.cid]   -> ', !modelMap[existing.$.cid])

          if (!modelMap[existing.$.cid]) {
            modelMap[existing.$.cid] = true;
            set.push(existing);
          }

          models[i] = existing;
        }

        else if (add) {

          model = models[i] = this._prepareModel(model, options);

          if (model) {
            toAdd.push(model);
            this._addReference(model, options);
            modelMap[model.$.cid] = true;
            set.push(model);
          }

        }

      })

      if (remove) {
        this._removeModels(this.models.filter(model => undefined === modelMap[model.$.cid]), options);
      }

      // console.log('set.length', set.length)
      // console.log('replace', replace)


      if (set.length && replace) {
        this.$.models.length = 0;
        this.$.models.splice(0, 0, ...set);
      }

      else if (toAdd.length) {
        this.$.models.splice(this.$.models.length, 0, ...toAdd);
      }

      // console.log('<--- collection.set')
      // console.log('----------------------------------------------------------------')
      return singular ? models[0] : models;
    }

    modelId(attrs) {
      return attrs &&
        this.$.model &&
        new this.$.model(attrs).id ||
        undefined
    }

    // collection.get
    get(arg) {
      if (undefined === arg) return undefined;

      const $ = this.$;
      if (arg in $.byId) return $.byId[arg];
      if ('object' !== typeof (arg)) return undefined;
      if (!('$' in arg)) arg = { '$': arg };

      const id = this.modelId(arg.$.attributes || arg.$);
      return id && id in this.$.byId ?
        this.$.byId[id] :
        arg.$.cid && arg.$.cid in $.byId ? $.byId[arg.$.cid] : undefined;
    }

    has(arg) {
      return !!this.get(arg);
    }

    toJSON(options) {
      return this.$.models.map(model => model.toJSON(options));
    }


    parse(resp, opts) {
      return resp ? Sync.parse(resp) : resp;
    }

    sort(options = {}) {
      let comparator = this.comparator;

      if (!comparator) {
        throw new Error('é necessario um comparador para ordenar');
      }

      const length = comparator.length;
      if ('function' === typeof comparator) {
        comparator = comparator.bind.call(this);
      }

      if (length == 1 || 'string' === typeof comparator) {
        this.$.models = this.sortBy(comparator);
      }
      else {
        this.$.models.sort(comparator);
      }

      return this;
    }

    create(model, options) {
      options = clone(options);

      model = this._prepareModel(model, options);
      if (!model) return;

      this.add(model, options);

      return model.save(null, options);
    }

    fetch(options) {
      const collection = this;

      return Sync.fetch.call(collection, assign({ 'parse': true }, options))
        .then(({ attributes, response, options }) => {
          collection[options.reset ? 'reset' : 'set'](attributes, options);
          return { collection, response, options }
        });
    }

    _reset() {
      this.$.models = [];
      this.$.byId = {};
    }

    _isModel(model) {
      return model instanceof this.$.model
    }

    _prepareModel(attrs, options) {

      if (this._isModel(attrs)) {
        if (!(attrs.$.collection)) attrs.$.collection = this;
        return attrs;
      }

      return new this.$.model(attrs, clone(options, { 'collection': this }));
    }

    _removeModels(models, options) {
      const removed = [];

      models.map(model => this.get(model)).filter(a => a).some(model => {

        // console.log('-----------------')

        // console.log(model)
        // console.log($.models)
        // console.log($.models.length)
        // console.log($.models.indexOf(model))

        this.$.models.splice(this.$.models.indexOf(model), 1);

        // console.log($.models)
        // console.log($.models.length)

        // console.log('-----------------')

        removed.push(model);
        this._removeReference(model, options);
      })

      return removed;
    }

    _addReference(model, options) {
      if (!('$' in model)) return;

      const id = this.modelId(model.attributes);

      this.$.byId[model.$.cid] = model;
      if (id) this.$.byId[id] = model;
    }

    _removeReference(model, options) {
      const id = this.modelId(model.attributes);

      delete this.$.byId[model.$.cid];
      if (id) delete this.$.byId[id];
      if (this === model.$.collection) delete model.$.collection;
    }

  }

  return { Collection, Model, Sync }
})

